search-books/src/main/java/nix/education/service/response/elasticsearch/ElasticsearchResponseServiceImpl.java

```diff
@@ -0,0 +1,178 @@
+import lombok.RequiredArgsConstructor;
+import lombok.SneakyThrows;
+import nix.education.entity.ElasticBook;
+import nix.education.service.BookDataService;
+import nix.education.util.DateFormat;
+import nix.education.util.SearchConstants;
+import org.apache.commons.lang3.StringUtils;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Service;
+
+import static nix.education.util.DateFormat.DATE_FORMAT_BOOK_REQUEST;
+
+@Service
+@RequiredArgsConstructor
+public class ElasticsearchResponseServiceImpl implements ElasticsearchResponseService {
+
+    private final BookDataService bookDataService;
+
+    @Value("${elastic.index}")
+    String index;
+
+    @SneakyThrows
+    @Override
+    public Map<String, Object> prepareResponse(SearchResponse<ElasticBook> response) {
+        if (response.hits().hits().isEmpty()) {
+            return Map.of("message", "Sorry, any documents weren't found!");
+        }
+        Map<String, Object> responseBody = new LinkedHashMap<>();
+        responseBody.put("numFound", response.hits().hits().size());
+        responseBody.put("docs", bookDataService.getBooks(getIdsArray(response)));
+        responseBody.put("facet_counts", prepareAggregations(response));
+
+        return responseBody;
+    }
+
+    private List<Long> getIdsArray(SearchResponse<ElasticBook> response) {
+        return response.hits().hits().stream()
+            .map(hit -> Long.parseLong(hit.id())).collect(Collectors.toList());
+    }
+
+    private Map<String, Object> prepareAggregations(SearchResponse<ElasticBook> response) {
+        List<Map<String, Object>> aggregations = new ArrayList<>();
+        response.aggregations()
+            .forEach((key, value) -> {
+                if (value.isSterms()) {
+                    aggregations.add(prepareTermsAggregations(key, value));
+                } else if (value.isAutoDateHistogram()) {
+                    aggregations.add(prepareRangeAggregations(key, value));
+                }
+            });
+        Map<String, Object> result = new HashMap<>();
+        aggregations.forEach(map -> map.keySet().forEach(key -> result.put(key, map.get(key))));
+        return result;
+    }
+
+    private Map<String, Object> prepareTermsAggregations(String name, Aggregate buckets) {
+        List<Object> buketList = new ArrayList<>();
+        for (StringTermsBucket bucket : buckets.sterms().buckets().array()) {
+            Map<String, Object> values = new HashMap<>();
+            values.put(SearchConstants.FACET_COUNT, bucket.docCount());
+            values.put(SearchConstants.FACET_LABEL, bucket.key()._get());
+            values.put(SearchConstants.FQ,
+                StringUtils.join(name, SearchConstants.COLON, bucket.key()._get()));
+            buketList.add(values);
+        }
+        return Map.of(name, buketList);
+    }
+
+    private Map<String, Object> prepareRangeAggregations(String name, Aggregate buckets) {
```

Comment:
Can we try to use date histogram here? I tried to play out with this a bit. This works at list thru the API:
![Screenshot from 2023-06-12 10-51-31](https://github.com/OlehNoskov/education-solr/assets/95301611/376a0067-075b-45bf-86ba-73da1e35356c)
I might assume that in the code this should look like:
![Screenshot from 2023-06-12 10-53-30](https://github.com/OlehNoskov/education-solr/assets/95301611/f77c998d-d6c3-451d-b5db-24fe43df29fc)
Please see this page for reference:
https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-bucket-datehistogram-aggregation.html